Monorepo setup

eyelidlessness opened the pull request:
## Notes on scope of PR

The following are considered **in scope** in this PR:

-   [x] Set up maintenance and development of all Enketo packages intended for either standalone or embedded deployment, or Enketo-maintained dependencies thereof.

-   [x] Minimal monorepo setup to facilitate said maintenance and development. Minimal as in:

    -   Tooling (Yarn for package management and running commands)
    -   CI (preserve pertinent CI functionality from each respective package, limit runtime and number of jobs where reasonable)
    -   Simple and idiomatic package/directory structure:
        -   Root README and requisite tooling configs
        -   All releasable packages under `/packages`
        -   Any extant or future demonstration packages under `/examples`

-   [x] Upgrade to currently supported versions of Node (18 and 20; this task had recently been deferred in anticipation of this monorepo migration)

-   [x] Upgrade direct (`package.json`) dependencies where either required for said Node upgrades, or appropriately unblocked by related tooling changes. This has mostly been limited to dev dependencies, with a few exceptions (most notably for runtime dependencies, upgrading jQuery was unblocked).

    One particular pain point worth calling out was handling changes to test snapshots in `enketo-transformer`. These changes were necessary because of changes in `vitest`'s snapshot format, `prettier`'s move to asynchronous-by-default formatting, and whitespace changes in `@prettier/plugin-xml`. Addressing the latter was the most important, without which there would be an absurd amount of false positive updates to the snapshot file. And as a future note, if we continue to maintain or expand snapshot tests we'll want to consider ways we can break them up to limit false positives in the future.

-   [x] Move management of dev dependencies to the monorepo root, as much as possible.

    Some exceptions were made for unshared dependencies which were fussy about their place in the file system relative to their respective dependent package.

-   [x] Consolidate code formatting (Prettier, ESLint) across all packages, now run as a single task at the monorepo root. There are of course quite a few _formatting_ exceptions (none new, just consolidated under `.eslintrc.json`).

    **Note:** code in the `openrosa-xpath-evaluator` package has been reformatted in this process. We had done a previous unification pass on the other packages, but deferred that one at the time because it was already fairly divergent in code style from the rest. But it made sense now to reconcile.

    This change, like the previous formatting pass, was **mostly automated**. The remaining manual changes were almost entirely so uncontroversial they're not really worth elaborating. The only real exception would be the changes from `isNaN` to `Number.isNaN`. The former coerces its argument to a number before checking its `NaN`-ness, which can have surprising behaviors (which, in turn, is why it's a lint error). I tried to be as conservative in addressing these errors as possible, but they're worth a close eye.

-   [x] Correct, and in some cases improve, module references between packages. Most notably, Sass imports should be much more clear.

-   [x] Any fixes necessary to build and dev for each package. Many of these were minor and should hopefully be obvious, significant changes in `enketo-express`'s `Gruntfile.js` are worth a close look.

-   [x] All packages are built on install. In some cases this ensures inter-package dependencies work as expected without additional steps. In other cases a stub `build` step was introduced for consistency. (The latter wasn't strictly necessary, but helped to ensure correct top-level behavior.)

-   [x] Consolidate GitHub issue/PR templates. (Note: I assume this worked, but it won't be testable until they land in `main`)

-   [x] Oh yeah, the main branch is now `main`

The following **may be** considered in scope for this PR (or a later change):

-   [ ] Remove now-unused Git history migration scripts. They have served their purpose, we can leave some breadcrumbs back to them in the new Git commit history as appropriate.
-   [ ] Remove any temporary TODOs and notes as we see fit
-   [ ] Rebase as desired
-   [ ] Reduce redundancies in CI config. There's **quite a lot** of stuff repeated almost-or-totally-verbatim in `ci.yml`. In theory some of it can be broken out into separate actions which might be triggered explicitly by the main config. This sounds nice, but a word of caution: in past experience, I've found attempts to apply more than the minimal necessary abstraction in these sorts of configs (and in YAML broadly) _can and even tend to do more harm than good_.

The following are **in scope** for one or more future PRs:

-   [ ] Updated release automation and procedures. The current release processes are currently copied from their respective packages into `.github/workflows/ci.yml`, and commented out as stubs.
-   [ ] Update other packaging/deployment considerations (i.e. Dockerfile)
-   [ ] Additional revisions to package-level READMEs and documentation revisions

The following will be done when we determine the monorepo migration is sufficiently ready for switchover:

-   [ ] Migrate existing issues from their individual repos
-   [ ] Archive those repos

The following **may be** in scope for future work:

-   [ ] Migrate any VSCode `launch.json`/`tasks.json` as we see fit. (I suspect I was the only one using these, and I haven't used them much for quite a while. I'm inclined to reconsider the approach if/when we find a need.)
-   [ ] Consolidate top level `tsconfig.json` and `jsconfig.json` if/where possible. Some cases must remain broken out in separate configs (generally test-specifics, and cases like those in `enketo-transformer` with conditional access to DOM APIs). I would like to have spent some time addressing this in the initial setup PR, but there's some inherent complexity (i.e. [composite](https://www.typescriptlang.org/tsconfig#composite) [projects](https://www.typescriptlang.org/docs/handbook/project-references.html)) that should probably be addressed, and a bunch of incidental complexity related to our intended gradual adoption.
-   [ ] Improve consistency of scripts and task management (i.e. `package.json` `scripts` entries, remaining usage of `grunt`). I'd like to have done more on this here too, but it felt increasingly like a separate scope as the rest progressed.
-   [ ] We should eventually address issues with `enketo-transformer`'s release-related portions of `package.json` found by [Are the types wrong?](https://arethetypeswrong.github.io/?p=enketo-transformer%403.0.1).
-   [ ] TODOs extracted from `.gitignore`:
    -   [ ] Allow `*.code-workspace` if we ever want that
    -   [ ] Naming and path consistency of build artifacts, don't mix build output in source paths
    -   [ ] Naming and path consistency of fixtures
    -   [ ] Consolidate configs
    -   [ ] Revisit Docker/etc related ignores

## Notes on other Enketo repositories

There are quite a few other Enketo repositories, some of which may be considered for inclusion in this monrepo. It's taken as read that any archived repository would not be under consideration. Some quick thoughts on the rest:

-   `enketo-core-performance-monitor`: I'd like to bring this in, and address some other issues with it in the process.
-   `vex-dialog-enketo`: this is an explicit dependency of `enketo-express` (and a fork of `vex-dialog`). It could make sense to include as long as we continue using it. But it's also been remarkably stable (i.e. last commit was in 2018), so I'm currently inclined to leave it that way.
-   `merge-xml`: also a fork, and presently involved in at least one known bug (restoring non-relevant subtrees of repeat instances in submissions). It very probably makes sense to bring in, or revisit its usage.
-   `Leaflet.draw`: also a fork, also remarkably stable. Inclined to leave out as well.
-   `blog`, `enketo-website`, `enketo-roadmap`, `enketo-api-docs`: these would potentially make sense to include, but we should evaluate their maintenance generally.
-   `enketo-validate`, `enketo-validate-webapp`, (maybe) `centro`: I wasn't even aware these were being actively maintained until I looked today! To the extent they are a going concern, it seems a case could be made for their inclusion.
-   `enketo-iframe-demo`: could be reasonably migrated into `/examples`.

## Breaking changes

-   All packages: Dropped support for Node 12, 14, 16. Versions now supported: 18 and 20.
-   `enketo-express`: Removed support for `request-filtering-agent` and related configs (`ip filtering`). We believe this support is no longer necessary following previous changes to how manifest media is served.

## Risks

_Overall_ this **should** be relatively low risk. Some specific risks and considerations are discussed above. Beyond those:

-   Yarn does not have an equivalent to `npm audit fix [--production]`, and the corresponding release instructions have been removed from packages' respective READMEs. Yarn **does** provide an `audit` command which we will continue to use, so really the risk here is an increased emphasis on manually resolving dependency vulnerabilities. On the other hand, Yarn's `resolutions` feature gives us a much freer hand to address said vulnerabilities. In past experience, I haven't found the lack of a `fix` command to be an issue at all.

-   Quite a lot of the version updates are potential risk factors. I've taken care to read through release notes to evaluate all explicit breaking changes, and believe each upgrade was safe according to the information available to me.

    It'll be particularly worth double checking that the Sass upgrades haven't introduced meaningful changes to built stylesheets.

-   Changing development process can be disruptive and might trip up contributors. In my experience, the expectation of working from a monorepo's root can be especially surprising.

-   I tried to pay close attention to these changes in general, and particularly whenever anything package-local was consolidated at the top level. One area of risk in that vein I think is worth pointing out is the possibility that any safeguards have been **mistakenly relaxed**, e.g. unintentional loosening of lint rules. I don't think we can categorically rule out that possibility, though I think the changes generally demonstrate tightening of rules if anything.

<!--
## Inline TODOs

Some TODOs were added inline which might be worth a look.

-   [ ] `.gitignore`
    -   [ ] Notes about file path consistency (fixture paths, build output). There's a general story about filesystem consistency which might warrant a separate issue.
    -   [ ] Note about config consistency. Yesterday we discussed potentially beginning to address this along with forthcoming config changes around themes.

## Other notes

-   `.editorconfig` may be redundant? It has the nice property of automatically enforcing _some_ code style consistency in _some_ editors, but not much.

## Potential TODO

-   [ ] Rebase as needed

### Tooling

-   [ ] VSCode Launch commands (test/debug modes).

    These were not included, primarily because I believe I was probably the only person using them... and I haven't used them much for a while. If there's a desire to adapt them to the monorepo, we can do so, otherwise maybe we remove them for now? -->

lognaturel reviewed (commented):
Phew, that's a lot of housekeeping! It looks very thoughtfully done to me and I don't see any additional sources of risk beyond what's listed in the PR description. I went through commit-by-commit thinking about any possible alternatives, rebasing to consider and risk.

I felt some mild irritation when doing `yarn install` that there seems to be less feedback about what's going on than there was with `npm i`. In particular, I got no console output while `libxslt` was building which made me think things were broken. Maybe that's something that can be configured and could be considered in the future. Other than that, everything worked smoothly and feels very coherent. ✨ 

Do you want to consider any changes for the .gitignore before we merge? Or maybe there's some rebasing you want to do? I don't see any particular need.

lognaturel commented on .gitignore:
```diff
@@ -0,0 +1,165 @@
+
+# Introduced in 1a5dc306cd1c741290fcd7b9973eb3bd1001cb40.
+# Drafts... of CHANGELOG? Some other prose related to the release?
+/packages/enketo-express/**/_drafts
```
I think we should just remove this and it can be reintroduced if found to be needed.

eyelidlessness commented:
> I felt some mild irritation when doing yarn install that there seems to be less feedback about what's going on than there was with npm i. In particular, I got no console output while libxslt was building which made me think things were broken.

That tripped me up at first too. I think `node-sass` might even take longer than `libxslt` and `libxmljs` too, but you wouldn't necessarily notice it from the output. If you do want the output, you can run `yarn install --verbose`. Personally I came to appreciate the lack of output once I got used to the wait without it... but yeah I can see why it might seem broken.

> Do you want to consider any changes for the .gitignore before we merge? Or maybe there's some rebasing you want to do? I don't see any particular need.

I've removed the TODOs from there and added in the PR notes (for now). Up to you on whether you want to rebase any of that.

eyelidlessness commented on .github/workflows/ci.yml:
```diff
@@ -46,7 +46,7 @@ jobs:
         strategy:
             matrix:
-                node-version: [ '18', '20' ]
+                node-version: ['18.18.0', '20.5.1']
```
I think I looked into this and it wasn't going to work for some reason? "Variables" in GHA means environment variables, and I think they're not available as matrix input but I might be misremembering.

lognaturel replied:
Whyyyyy... https://stackoverflow.com/questions/74072206/github-actions-use-variables-in-matrix-definition
